composer/utils/checkpoint.py

```diff
@@ -200,46 +201,158 @@ def load_checkpoint(
+                # If fsdp_sharded_state_dict_enabled then just use that rank's unique tempdir.
+                node_checkpoint_folder = (tempdir if state.fsdp_sharded_state_dict_enabled else
+                                          _get_local_rank_zero_path(tempdir))
+                assert node_checkpoint_folder is not None
+
+                composer_states_filepath, extracted_checkpoint_folder, extracted_rank_n = download_checkpoint(
+                    path=path,
+                    node_checkpoint_folder=node_checkpoint_folder,
+                    object_store=object_store,
+                    progress_bar=progress_bar,
+                    fsdp_sharded_state_dict_enabled=state.fsdp_sharded_state_dict_enabled,
+                    deepspeed_sharded_checkpoint=is_model_deepspeed(state.model),
+                )
+                rng_state_dicts = _restore_checkpoint(
+                    state,
+                    logger,
+                    composer_states_filepath,
+                    extracted_rank_n,
+                    extracted_checkpoint_folder,
+                    load_weights_only=load_weights_only,
+                    strict_model_weights=strict_model_weights,
+                    ignore_keys=ignore_keys,
+                    exclude_algorithms=exclude_algorithms,
+                    algorithm_passes=algorithm_passes,
+                )
+            finally:
+                # Wait for all ranks to finish restoring the checkpoint before releasing the tempdir, since tempdir can
+                # be a shared resource between nodes.
+                dist.barrier()
+
+        log.info('%s loaded from %s', 'Model weights' if load_weights_only else 'Trainer checkpoint', path)
+    return rng_state_dicts
+
+
+def load_sharded_checkpoint(
+    source_path: str,
+    state: State,
+    logger: Logger,
+    object_store: Optional[Union[ObjectStore, LoggerDestination]] = None,
+    load_weights_only: bool = False,
+    strict_model_weights: bool = False,
+    progress_bar: bool = True,
+    ignore_keys: Optional[Union[List[str], Callable[[Dict], None]]] = None,
+    exclude_algorithms: Optional[List[str]] = None,
+    algorithm_passes: Optional[List[AlgorithmPass]] = None,
+) -> Dict:
+
+    if not using_torch_2():
+        raise ValueError(f'Sharded checkpoint loading requires torch version >= 2.0.0 Got {torch.__version__}')
+
+    from torch.distributed import checkpoint as dist_cp
+    from torch.distributed.checkpoint.metadata import Metadata
+    from torch.distributed.checkpoint.optimizer import load_sharded_optimizer_state_dict
+
+    # This function is used so we can figure out which ranks need to load saved rng's and which can just make their own.
+    def _get_num_ranks_that_saved_rng(metadata: Metadata):
+        rng_inds = []
+        for field_name, field_value in metadata.planner_data.items():
+            if 'rng' in field_name:
```

Comment:
what are the field names? is there a reason this check is `in` rather than `==`? I assume yes, just wondering if we can make the check more explicit/looking to understand what is in this `metadata.planner_data` dict

Reply:
rng state is a list and field_names are FQN's and they [append the index value to each item in a list](https://github.com/pytorch/pytorch/blob/28f43c767c0de48885f8de285ca4af4424d28f29/torch/distributed/checkpoint/_traverse.py#L72-L74), so this means the each rank's rng will be an fqn with 'rng' somewhere in there along with the rank. Seems easier just to check 'rng' in `field_name`